### DISCUSSION TITLE
PartTree question

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#77] [Q&A] [answered]
https://github.com/mjhoptics/ray-optics/blob/1d1624115607fcd7fc2d157b3fb4141bc8efc1eb/src/rayoptics/elem/parttree.py#L262

In the above code, the variable i is defined in the for loop, but also modified inside the loop.
Trying to understand the intention here.

**Comment by mjhoptics** [comment, marked as answer]
This is code that I hope I never have to change again... The use cases it's attempting to handle include:

1. single lens
2. single mirror
3. cemented lenses (2 or more)
4. Mangin mirror (lens with mirror coat on rear surface, i.e. a buried reflector)
5. cemented Mangin mirror

The modification of *i* within the loop is intended and there are no side effects because every time thru the loop, the enumerate function returns a new value for *i*. I can't say that I'm proud of this code, but it works.

**Reply by dibyendumajumdar** [reply]
Okay thanks; it is a complex piece of logic there.
